### ISSUE TITLE
Hackage's search and package pages are no longer accessible

### ISSUE THREAD

**Issue by Merivuokko** [haskell/hackage-server#1051]
A recent update broke Hackage's accessibility. Search and package pages no longer work with browsers which do not have support for JavaScript execution. Such browsers include Lynx and Elinks, which are commonly used by blind users.

Possible workarounds include: 1) provide the original content of these pages within a <nosciprt/> section, and 2) provide separate accessible pages for non-JS browsers.

On Unix-derivatives blind users are not able to switch over to JS-supporting browsers, as the graphical screen reading software on these platforms is inferior.

To date hackage has been an exceptionally accessible site, and I wish to see this support continued.

**Comment by gbaz** [comment]
cc: @ysangkok

**Comment by Merivuokko** [comment]
I tested the new pages with Lynx (I previously used Elinks) and I found out, that the package pages work with Lynx. So with package pages the issue is something else than JS – probably a redirection issue.

If I direct Elinks to https://hackage.haskell.org/package/containers, I et the following kind of JSON as a response:

```json
{"0.1.0.0":"normal","0.1.0.1":"normal","0.2.0.0":"normal","0.2.0.1":"normal","0.3.0.0":"normal", ...}
```

If I load the same URL with Lynx, I get a warning:

```
Alert!: HREF in BASE tag is not an absolute URL.
```

But then the page loads fine.

**Comment by gbaz** [comment]
I think the former issue (elinks) is because there is now a json content type and hackage server will default to it _in this specific instance_ if a browser does not specify a desired content type. I find it weird that elinks doesn't send a content-type tag, but perhaps this should encourage us to enforce that the "default content type" is always text/html if it exists.

Not sure about the base tag issue -- prs welcome.

Vis a vis putting the old code on alternative endpoints -- I suppose this is ok. I would much prefer we had a path that enabled new search but computing the results on the server side rather than the browser, but I understand if that would be rather too much work...

**Comment by ysangkok** [comment]
The HTML versions of the package pages did not have any JavaScript added, in fact a bit of JavaScript was removed as part of #1022. But you are correct that the browse and search pages now require JavaScript. This could be restored as previously mentioned.

Here is how to reproduce the default content-type issue pointed out by @gbaz:

```
$ curl -s -D - "https://hackage.haskell.org/package/containers" | grep -i Content-type
Content-Type: application/json
$ curl -s -D - -H "Accept: text/html" "https://hackage.haskell.org/package/containers" | grep -i Content-type
Content-Type: text/html; charset=utf-8
```

This is a bit curious since the recent PR #996 by @Kleidukos was meant to default to text/html: [relevant code change](https://github.com/haskell/hackage-server/pull/996/files#diff-e472f8254e83d4ed34adcac7f603ab69eda10c76706a2e3b36fa27cbde675d9eR524). But really I think that browsers that try to parse HTML should send a header that signals that.

**Comment by ysangkok** [comment]
Ok, I have written some code for this that I will submit after #1049 is in. Here it is: [diff](https://github.com/ysangkok/hackage-server/compare/janus/faster-index...ysangkok:janus/noscript?expand=1).

**Comment by gbaz** [comment]
merged the above pr

**Comment by gbaz** [comment]
Also would appreciate if you could figure out why the content-type fix doesn't work, though as you say it shouldn't be an issue for basically any modern browser.

**Comment by ysangkok** [comment]
Sure! I'll make sure Elinks support is improved. I do think it would be appropriate to have two different PRs for these fixes. I'll get the noscript search PR out tonight, so in about 8 hours.

**Comment by ysangkok** [comment]
@Merivuokko A new Elinks-friendly page is now available at https://hackage.haskell.org/packages/noscript-search . Does it work for you?

Here is the documentation for the query language that is available:

```
(downloads > 1000)
    Only show packages with more than 1000 downloads within the last 30 days. The download count is inexact because Hackage uses a [content delivery network](https://en.wikipedia.org/wiki/Content_delivery_network).
(lastUpload < 2021-10-29)
    Only show packages for which the last upload was before (i.e. excluding) the given UTC date in [the 'complete date' format as specified using ISO 8601](https://www.w3.org/TR/NOTE-datetime).
(lastUpload = 2021-10-29)
    Only show packages for which the last upload was within the 24 hours of the given UTC date.
(maintainer:SimonMarlow)
    Only show packages for which the maintainers list includes the user name [SimonMarlow](https://hackage.haskell.org/user/SimonMarlow).
(tag:bsd3)
    Only show packages with the [bsd3](https://hackage.haskell.org/packages/tag/bsd3) tag.
(not tag:network)
    Do not show packages with the [network](https://hackage.haskell.org/packages/tag/network) tag. The not operator can also be used with other filters.
(ageOfLastUpload > 5d)
    Only show packages uploaded more than five days ago.
(ageOfLastUpload > 4w)
    Only show packages uploaded more than four weeks ago. A week has seven days.
(ageOfLastUpload < 1m)
    Only show packages last uploaded less than one month ago. A month is considered to have 30.437 days.
(ageOfLastUpload < 2.5y)
    Only show packages last uploaded less than 2.5 years ago. A year is considered to be 365.25 days.
(rating > 2.5)
    Only show packages with a rating of more than 2.5. The dot is the only accepted decimal separator.
(rating /= 0)
    Only show packages with a rating unequal to zero.
(deprecated:any)
    Do not filter out deprecated packages. This must be explicitly added if desired.
(deprecated:true)
    Only show deprecated packages.
(deprecated:false)
    Only show packages that are not deprecated. If no other deprecation filter is given, this filter is automatically added.
(distro:Debian)
    Only show packages that are available in the Debian distribution. See the [full list of available distributions](https://hackage.haskell.org/distros).
```

**Comment by Merivuokko** [comment]
On 2022-05-15 at 09:01 -0700, Janus Troelsen ***@***.***> wrote:
 > @Merivuokko A new Elinks-friendly page is now available at https://hackage.haskell.org/packages/noscript-search . Does it work for you?

Yes, mostly. I tried to find out a syntax to combine predicates, but could not
find any. The page does not tell me if I have a syntax error in my query.

I got a HTTP 502 once, with this query:
`(downloads > 1000) && (ageOfLastUpload < 3m)`
but it does not fail any more, just no results.

-- 
Aura

**Comment by ysangkok** [comment]
@Merivuokko If multiple predicates are found, they are all applied. So there are no logical operators, and there is no option to express anything except the implicit "and" operator. This is inspired by e.g. Google's query language, where if you use e.g. `intitle:index.of` and `site:.gov`, both of these filters will be applied. I am not sure how best to document this, especially since the documentation currently is only shown on the JavaScript search page...

Anything that doesn't look like a predicate will be used as a regular old full-text search term. So there is no feedback mechanism for notifying the user if something can't be parsed as a predicate.

**Comment by Merivuokko** [comment]
On 2022-05-15 at 14:46 -0700, Janus Troelsen ***@***.***> wrote:
 > @Merivuokko If multiple predicates are found, they are all applied. So
 > there are no logical operators, and there is no option to express anything
 > except the implicit "and" operator. This is inspired by e.g. Google's query
 > language, where if you use e.g. `intitle:index.of` and `site:.gov`, both of
 > these filters will be applied. I am not sure how best to document this,
 > especially since the documentation currently is only shown on the
 > JavaScript search page...

In my opinion a simple link to a separate documentation page would be the
best. That way the docs don't clutter the interface.

-- 
Aura

**Comment by ysangkok** [comment]
Most websites don't have separate documentation. I am not opposing it, but it would be a departure from how it seems things are currently done. We would need to choose a consistent way of displaying it.

Another options is inline documentation, which is hidden by default (like on the JavaScript enabled search page). Some user interfaces have question marks in circles, which show help texts once you hover the mouse cursor over it or alternatively click it (to enable accessibilty, and to use with tablets that don't allow for hovering). This model could be applied to Hackage too, but it seems like a pretty big change to me. Many form pages are heavy in text explaining things, presumably all of that could be moved into this "help system". If there is consensus for this, we should make a separate issue.

**Comment by Merivuokko** [comment]
> Most websites don't have separate documentation.

As far as I understand, this javascriptless version will be used only by people who have a browser without JavaScript support. (Which nowadays is not very many.) If this one page deviates from web UI conventions, I don't see it as a big issue.

 > Another options is inline documentation, which is hidden by default (like on the JavaScript enabled search page). Some user interfaces have question marks in circles, which show help texts once you hover the mouse cursor over it or alternatively click it (to enable accessibilty, and to use with tablets that don't allow for hovering). This model could be applied to Hackage too, but it seems like a pretty big change to me. Many form pages are heavy in text explaining things, presumably all of that could be moved into this "help system". If there is consensus for this, we should make a separate issue. 

I'm not aware of any way of doing this without JavaScript (and possible CSS) neither of which is really supported in terminal browsers.

-- 
Aura
